src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,267 @@ PhaseStatus Compiler::optUnrollLoops()
 #pragma warning(pop)
 #endif
 
+//-----------------------------------------------------------------------------
+// optIfConvert
+//
+// Find blocks representing simple if statements represented by conditional jumps
+// over another block. Try to replace the jumps by use of conditional nodes.
+//
+// Arguments:
+//   block -- block that may represent the conditional jump in an if statement.
+//
+// Returns:
+//   true if any IR changes possibly made.
+//
+bool Compiler::optIfConvert(BasicBlock* block)
+{
+#ifndef TARGET_ARM64
+    return false;
+#else
+
+    // Don't optimise the block if it is inside a loop
+    // When inside a loop, branches are quicker than selects.
+    // Detect via the block weight as that will be high when inside a loop.
+    if (block->getBBWeight(this) > BB_UNITY_WEIGHT)
+    {
+        return false;
+    }
+
+    // Does the block end by branching via a JTRUE after a compare?
+    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    {
+        return false;
+    }
+
+    // Verify the test block ends with a conditional that we can manipulate.
+    GenTree* last = block->lastStmt()->GetRootNode();
+    noway_assert(last->gtOper == GT_JTRUE);
+    if (!last->AsOp()->gtOp1->OperIsCmpCompare() || (last->gtFlags & GTF_SIDE_EFFECT) != 0 ||
+        (last->AsOp()->gtOp1->gtFlags & GTF_COLON_COND) != 0)
+    {
+        return false;
+    }
+
+    BasicBlock* middle_block = block;
+    GenTree*    asg_node     = nullptr;
+    Statement*  asg_stmt     = nullptr;
+    bool        found_select = false;
+
+    // Check the the block is followed by a block or chain of blocks that only contain NOPs and
+    // a single ASG statement. The destination of the final block must point to the same as the
+    // true path of the JTRUE block.
+    bool found_middle = false;
+    while (!found_middle)
+    {
+        middle_block = middle_block->bbNext;
+        noway_assert(middle_block != nullptr);
+
+        if (middle_block->bbNext == block->bbJumpDest)
+        {
+            // This is our final middle block.
+            found_middle = true;
+        }
+
+        // Make sure there is only one block which jumps to the middle block,
+        // and the middle block is not the start of a TRY block or an exception handler.
+        noway_assert(!fgCheapPredsValid);
+        if (middle_block->NumSucc() != 1 || middle_block->bbJumpKind != BBJ_NONE ||
```

Comment:
> the middle block is not the start of a TRY block or an exception handler

The correct way to handle this would be:

```
bbIsTryBeg(middle_block) || bbIsHandlerBeg(middle_block)
```